File: packages/vitals_domain/lib/src/record.dart
import 'fact.dart';

/// A patient's record: the set of its facts, ordered by stamp. Nothing here
/// deletes. What a reader sees is [current]: every fact that has not been
/// superseded, with the superseded ones one step away in [history].
final class Record {
  Record._(this.patient, Map<String, Fact> facts) : _facts = facts;

  factory Record.empty(List<int> patient) => Record._(patient, {});

  factory Record.of(List<int> patient, Iterable<Fact> facts) {
    final r = Record.empty(patient);
    for (final f in facts) {
      r._facts[f.key] = f;
    }
    return r;
  }

  final List<int> patient;
  final Map<String, Fact> _facts;

  /// Every fact ever recorded, in stamp order.
  List<Fact> get all => _facts.values.toList()..sort();

  int get length => _facts.length;

  bool contains(Fact f) => _facts.containsKey(f.key);

  /// The facts a reader is shown: not superseded, and not themselves
  /// supersessions.
  List<Fact> get current {
    final superseded = <String>{
      for (final f in _facts.values)
        if (f.kind == FactKind.supersession && f.supersedes != null)
          _key(f.supersedes!),
    };
    return all
        .where((f) =>
            f.kind != FactKind.supersession && !superseded.contains(f.key))
        .toList();
  }

  /// Everything that corrected a fact, and everything that corrected those,
  /// in stamp order. Two devices may correct the same fact while apart —
  /// both corrections survive, and both are here; the record shows neither
  /// as hidden and the reader decides, which is the nurse's call, not the
  /// merge's.
  List<Fact> history(Fact f) {
    final out = <Fact>[];
    var frontier = <String>{f.key};
    while (frontier.isNotEmpty) {
      final next = <String>{};
      for (final s in all) {
        if (s.kind == FactKind.supersession &&
            s.supersedes != null &&
            frontier.contains(_key(s.supersedes!))) {
          out.add(s);
          next.add(s.key);
        }
      }
      frontier = next;
    }
    return out;
  }

  static String _key(List<int> id) {
    const digits = '0123456789abcdef';
    final out = StringBuffer();
    for (final b in id) {
      out.write(digits[b >> 4]);
      out.write(digits[b & 0xf]);
    }
    return out.toString();
  }
}

File: server/src/Vitals.Domain/Record.cs
namespace Vitals.Domain;

/// <summary>A patient's record: the set of its facts. Nothing here deletes.</summary>
public sealed class Record
{
    private readonly Dictionary<string, Fact> _facts;

    private Record(byte[] patient, Dictionary<string, Fact> facts)
    {
        Patient = patient;
        _facts = facts;
    }

    public byte[] Patient { get; }

    public static Record Empty(byte[] patient) => new(patient, new Dictionary<string, Fact>(StringComparer.Ordinal));

    public static Record Of(byte[] patient, IEnumerable<Fact> facts)
    {
        var r = Empty(patient);
        foreach (var f in facts) r._facts[f.Key] = f;
        return r;
    }

    public int Length => _facts.Count;
    public bool Contains(Fact f) => _facts.ContainsKey(f.Key);

    /// <summary>Every fact ever recorded, in stamp order.</summary>
    public List<Fact> All
    {
        get
        {
            var list = _facts.Values.ToList();
            list.Sort();
            return list;
        }
    }

    /// <summary>The facts a reader is shown: not superseded, and not themselves supersessions.</summary>
    public List<Fact> Current
    {
        get
        {
            var superseded = new HashSet<string>(StringComparer.Ordinal);
            foreach (var f in _facts.Values)
            {
                if (f.Kind == FactKind.Supersession && f.Supersedes is not null) superseded.Add(Convert.ToHexStringLower(f.Supersedes));
            }
            return All.Where(f => f.Kind != FactKind.Supersession && !superseded.Contains(f.Key)).ToList();
        }
    }
}

/// <summary>Set union. The whole merge, by construction commutative, associative and idempotent.</summary>
public static class Merge
{
    public static Record Union(Record a, Record b)
    {
        if (!Bytes.Same(a.Patient, b.Patient)) throw new ArgumentException("records of two patients cannot merge");
        return Record.Of(a.Patient, a.All.Concat(b.All));
    }
}
